File: src/types/GameTypes.ts
export type GameStatus = 'playing' | 'won' | 'draw';
export type PlayerType = 'human' | 'ai';
export type GameMode = 'singleplayer' | 'multiplayer' | 'tournament' | 'practice';

export interface PlayerStats {
    totalGames: number;
    wins: number;
    losses: number;
    draws: number;
    winRate: number;
    winStreak: number;
    currentStreak: number;
    averageMovesPerGame: number;
    totalMoves: number;
    averageGameLength: number;
    favoriteMoves: Array<{
        position: string;
        count: number;
    }>;
    winningPatterns: Array<{
        pattern: string;
        count: number;
    }>;
    performanceHistory: Array<{
        date: string;
        result: 'win' | 'loss' | 'draw';
        gameLength: number;
    }>;
}

export interface Player {
    id: string;
    name: string;
    type: PlayerType;
    symbol: 1 | 2;
    aiConfig?: AIConfig;
    stats: PlayerStats;
}

export interface Move {
    player: 1 | 2;
    position: [number, number];
    timestamp: number;
}

export interface GameState {
    board: number[][];
    currentPlayer: 1 | 2;
    gameStatus: GameStatus;
    winner: 1 | 2 | null;
    players: [Player, Player];
    moveHistory: Move[];
    undoStack: Move[];
    redoStack: Move[];
    winningCells?: Array<[number, number]>;
    sessionStats: {
        totalGames: number;
        startTime: number;
        lastUpdateTime: number;
    };
}

export interface AIConfig {
    difficulty: 'easy' | 'medium' | 'hard';
    thinkingTime?: number;
}

export interface TournamentSettings {
    rounds: number;
    timeLimit: number;
    winningScore: number;
}

File: PoTicTac.Client/Models/GameTypes.cs
namespace PoTicTac.Client.Models
{
    public enum GameStatus
    {
        Playing,
        Won,
        Draw
    }

    public enum PlayerType
    {
        X,
        O,
        None,
        Draw,
        Human,
        AI
    }

    public enum GameMode
    {
        Singleplayer,
        Multiplayer,
        Tournament,
        Practice
    }

    public enum Difficulty
    {
        Easy,
        Medium,
        Hard
    }

    public class PlayerStats
    {
        public int TotalGames { get; set; }
        public int Wins { get; set; }
        public int Losses { get; set; }
        public int Draws { get; set; }
        public double WinRate { get; set; }
        public int WinStreak { get; set; }
        public int CurrentStreak { get; set; }
        public double AverageMovesPerGame { get; set; }
        public int TotalMoves { get; set; }
        public double AverageGameLength { get; set; }
        public List<FavoriteMove> FavoriteMoves { get; set; } = new();
        public List<WinningPattern> WinningPatterns { get; set; } = new();
        public List<PerformanceRecord> PerformanceHistory { get; set; } = new();
    }

    public class FavoriteMove
    {
        public string Position { get; set; } = string.Empty;
        public int Count { get; set; }
    }

    public class WinningPattern
    {
        public string Pattern { get; set; } = string.Empty;
        public int Count { get; set; }
    }

    public class PerformanceRecord
    {
        public DateTime Date { get; set; }
        public string Result { get; set; } = string.Empty; // "win", "loss", "draw"
        public int GameLength { get; set; }
    }

    public class Player
    {
        public string Id { get; set; } = string.Empty;
        public string Name { get; set; } = string.Empty;
        public PlayerType Type { get; set; }
        public PlayerType Symbol { get; set; } // X or O
        public AIConfig? AiConfig { get; set; }
        public PlayerStats Stats { get; set; } = new();
    }

    public class Move
    {
        public PlayerType Player { get; set; }
        public int[] Position { get; set; } = new int[2]; // [row, col]
        public long Timestamp { get; set; }
    }

    public class GameBoardState
    {
        public const int BoardSize = 6;

        public PlayerType[,] Board { get; set; } = new PlayerType[BoardSize, BoardSize];

        public GameBoardState()
        {
            // Initialize all cells to None
            for (int i = 0; i < BoardSize; i++)
            {
                for (int j = 0; j < BoardSize; j++)
                {
                    Board[i, j] = PlayerType.None;
                }
            }
        }

        public PlayerType[,] Cells => Board; // Backward compatibility
        public PlayerType CurrentPlayer { get; set; }
        public GameStatus GameStatus { get; set; }
        public PlayerType? Winner { get; set; }
        public Player[] Players { get; set; } = new Player[2];
        public List<Move> MoveHistory { get; set; } = new();
        public List<Move> UndoStack { get; set; } = new();
        public List<Move> RedoStack { get; set; } = new();
        public List<int[]>? WinningCells { get; set; }
        public SessionStats SessionStats { get; set; } = new();
    }

    public class SessionStats
    {
        public int TotalGames { get; set; }
        public long StartTime { get; set; }
        public long LastUpdateTime { get; set; }
    }

    public class AIConfig
    {
        public Difficulty Difficulty { get; set; }
        public int? ThinkingTime { get; set; }
    }

    public class TournamentSettings
    {
        public int Rounds { get; set; }
        public int TimeLimit { get; set; }
        public int WinningScore { get; set; }
    }

    public class MultiplayerGameInfo
    {
        public string GameId { get; set; } = string.Empty;
        public bool IsHost { get; set; }
    }

    public class GameScore
    {
        public int Player { get; set; }
        public int AI { get; set; }
    }
}
